crates/ibc/src/clients/ics07_tendermint/misbehaviour.rs

```diff
@@ -1,21 +1,82 @@
 use crate::prelude::*;
 
+use bytes::Buf;
+use ibc_proto::google::protobuf::Any;
 use ibc_proto::ibc::lightclients::tendermint::v1::Misbehaviour as RawMisbehaviour;
 use ibc_proto::protobuf::Protobuf;
+use prost::Message;
 use serde::{Deserialize, Serialize};
+use tendermint_light_client_verifier::ProdVerifier;
 
-use crate::clients::ics07_tendermint::error::Error;
+use crate::clients::ics07_tendermint::error::{Error, IntoResult};
 use crate::clients::ics07_tendermint::header::Header;
-use crate::core::ics24_host::identifier::ClientId;
+use crate::core::ics02_client::error::Error as Ics02Error;
+use crate::core::ics24_host::identifier::{ChainId, ClientId};
 use crate::Height;
 
-pub const TENDERMINT_MISBEHAVIOR_TYPE_URL: &str = "/ibc.lightclients.tendermint.v1.Misbehaviour";
+pub const TENDERMINT_MISBEHAVIOUR_TYPE_URL: &str = "/ibc.lightclients.tendermint.v1.Misbehaviour";
 
 #[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
 pub struct Misbehaviour {
-    pub client_id: ClientId,
-    pub header1: Header,
-    pub header2: Header,
+    client_id: ClientId,
+    header1: Header,
+    header2: Header,
+}
+
+impl Misbehaviour {
+    pub fn new(client_id: ClientId, header1: Header, header2: Header) -> Result<Self, Error> {
+        if header1.signed_header.header.chain_id != header2.signed_header.header.chain_id {
+            return Err(Error::invalid_raw_misbehaviour(
+                "headers must have identical chain_ids".into(),
+            ));
+        }
+
+        if header1.height() < header2.height() {
+            return Err(Error::invalid_raw_misbehaviour(format!(
+                "headers1 height is less than header2 height ({} < {})",
+                header1.height(),
+                header2.height()
+            )));
+        }
+
+        let untrusted_state_1 = header1.as_untrusted_block_state();
+        let untrusted_state_2 = header2.as_untrusted_block_state();
+
+        let verifier = ProdVerifier::default();
+
+        verifier.validate(&untrusted_state_1).into_result()?;
```

Comment:
I think `validate()` API name here is very vague. Maybe add a comment and/or change to a more verbose name (in the tendermint-rs PR).
In general it is not clear to me what are the differences among `validate_...`, `verify_...`, `check_...` and if we are consistent in naming.

Reply:
I used the term `validate()` for stateless input validations/checks that can be run on raw (proto) types e.g. header. But I agree that the names are not consistent. How about `verify_untrusted()` instead of `validate()`?

Reply:
or maybe `verify_validator_sets()`?

Reply:
Done on tendermint-rs side. Will update here once https://github.com/informalsystems/tendermint-rs/pull/1226 is merged and released.

Reply:
https://github.com/cosmos/ibc-rs/pull/215/commits/dc8dbc6e66c5fb15d8a7a631447ec3079576fc17